shielder/contract/lib.rs

```diff
@@ -0,0 +1,93 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+mod merkle;
+mod psp22;
+mod types;
+mod errors;
+mod mocked_zk;
+
+#[ink::contract]
+mod contract {
+
+    use ink::storage::Mapping;
+
+    use crate::{merkle::MerkleTree, psp22::PSP22, types::{Set, Scalar}, errors::ShielderError, mocked_zk::{ZkProof, self}};
+
+    #[derive(Clone, Copy, scale::Encode, scale::Decode)]
+    #[cfg_attr(
+        feature = "std", 
+        derive(scale_info::TypeInfo)
+    )]
+    pub enum OpPub {
+        Deposit(u128, AccountId, AccountId),
+        Withdraw(u128, AccountId, AccountId)
+    }
+
+    #[ink(storage)]
+    pub struct Contract {
+        nullifier_set: Set<Scalar>,
+        notes: MerkleTree,
+    }
+
+    impl Contract {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {
+                nullifier_set: Mapping::new(),
+                notes: MerkleTree::new()
+            }
+        }
+
+        #[ink(message)]
+        pub fn add_note(&mut self, op_pub: OpPub, h_note_new: Scalar) -> Result<(), ShielderError>{
+            self.process_operation(op_pub)?;
+            self.notes.add_leaf(h_note_new)?;
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn update_note(
+            &mut self,
+            op_pub: OpPub,
+            h_note_new: Scalar,
+            merkle_root: Scalar,
+            nullifier_old: Scalar,
+            proof: ZkProof,
+        ) -> Result<(), ShielderError> {
+            self.process_operation(op_pub)?;
+            self.notes.is_historical_root(merkle_root)?;
+            (!self.nullifier_set
```

Comment:
I think it would be cleanest to have a separate function `fn nullify(&mut self, x: Scalar) -> Result<(), ...>` that adds to the nullifier or outputs error, if the element is there already. This way you can replace this, and also remove the line `self.nullifier_set.insert(nullifier_old, &());`.